### ISSUE TITLE
VIA C3 Samuel 2 wrong L2 cache size detection

### ISSUE THREAD

**Issue by nickos222** [memtest86plus/memtest86plus#339]
Hello guys,
I believe there is a bug in the L2 cache size detection logic.
Memtest86 v6.20 detects 16MB when in reality the L2 cache size is 64kb.

Older (ancient) memtest versions detect the l2 size correctly:
Memtest-86 v3.5b
Memtest86+ v4.20

I had some old floppy disks with these earlier versions and decided to see if they still worked.

**Comment by x86fr** [comment]
Can you please confirm the CPUID of your CPU?

**Comment by nickos222** [comment]
Here you are:

vendor_id	: CentaurHauls
cpu family	: 6
model		: 7
model name	: VIA Samuel 2
stepping	: 1

Almost no linux 586 image can boot with this cpu anymore!
It is very difficult to get additional info for now, nothing really works decently or at all with this cpu and with 512mb ram.
I have tried many distros and most failed with oops or ileggal instruction messages.
Something is broken in recent linux kernel for this cpu.

**Comment by debrouxl** [comment]
For years, both the kernel and the user-space in most 32-bit x86 Linux distros have been built for 686-class computers, which assumes support for instructions such as the CMOV family. When running such software combinations on hardware which does not support CMOV (e.g. AMD K6-1, I tried that in the past), many illegal instructions occur, indeed.

**Comment by martinwhitaker** [comment]
Mageia still supports 586-class machines.

**Comment by nickos222** [comment]
According to VIA documentation the Extended CPUID L2 Cache Configuration encoding found on Samuel 2 differs from the Intel/AMD ones.

Intel/AMD use ECX bits 31:16 to report L2 Cache size in 1K units
VIA Samuel 2 uses ECX bits 31:24 to report L2 Cache size in 1K units and ECX bits 23:16 to report L2 associativity 

Memtest cpuid code erroneously takes into account the additional bits 23:16 for the l2 size and the size from 64kb becomes 16mb.

Memtest shall only use ECX bits 31:24 to report L2 Cache size on VIA C3 Samuel 2 processors (Family 6, Model 7, Stepping 0~7)

VIA C3 Ezra (Family 6, Model 7, Stepping 8+) and C3 Ezra-T (Family 6, Model 8, Stepping 0~7) processors may also use samuel2 l2 cache config encoding,
but I have neither the hardware nor any documentation to verify it.

VIA Nehemiah (Family 6, Model 9, Stepping 0+) probably needs no fix because it uses the Intel/AMD defined l2 cache config encoding.

**Comment by nickos222** [comment]
I shall also provide a register dump for cpuid 0x80000006:

0x80000006 0x00: eax=0x00000000 ebx=0x00000000 ecx=0x40040120 edx=0x00000000

**Comment by nickos222** [comment]
Hello again,
linux kernel applies a bit-shift on centaur family 6, model 7&8 processors (Samuel2,Ezra and EzraT).
It looks like ezra also uses the 8bit L2 Cache Configuration encoding.

	/* VIA C3 CPUs (670-68F) need further shifting. */
	if ((c->x86 == 6) && ((c->x86_model == 7) || (c->x86_model == 8)))
		size >>= 8;

**Comment by x86fr** [comment]
I will revert to hardcoded value as previously. That's why untested code paths on CPU detection PR must be rejected. This was an exception...

**Comment by jonathan-teh** [comment]
Looks like a bug in #259. Don't think it's a regression as the hardcoded values are only used on CPUs without a brand string and this Samuel 2 has one (as does the WinChip 2 and newer), perhaps test 6.00 to be sure?
This should fix it:
```diff
diff --git a/system/cpuinfo.c b/system/cpuinfo.c
index 26e3e02..14f0341 100644
--- a/system/cpuinfo.c
+++ b/system/cpuinfo.c
@@ -80,6 +80,10 @@ static void determine_cache_size()
         if (cpuid_info.version.family == 5 || cpuid_info.version.family == 6) {
             l1_cache = cpuid_info.cache_info.l1_d_size;
             l2_cache = cpuid_info.cache_info.l2_size;
+            if (cpuid_info.version.family == 6 && (cpuid_info.version.model == 7 || cpuid_info.version.model == 8)) {
+                // Samuel 2/Ezra/Ezra-T
+                l2_cache >>= 8;
+            }
             break;
         } else if (cpuid_info.version.family != 7) {
             break;
```

**Comment by nickos222** [comment]
Jonathan's patch works just fine with memtest 6.20 .
Tested on a QDI A9 board, the initial tests where done on an MS6309.
Both of these barely support the c3, they misidentify the l2 cache.
ms6309 V2.1 reports L2 size 519KB
a9 reports L2 disabled

**Comment by nickos222** [comment]
Also I booted memtest v6.00 and it cannot detect any of the caches at all (N/A).

**Comment by x86fr** [comment]
Use `l2_cache = 512;` instead of `l2_cache >>= 8;`

**Comment by jonathan-teh** [comment]
> Use `l2_cache = 512;` instead of `l2_cache >>= 8;`

I think you mean `64`? I've changed the PR to hardcode the value. I'll be AFK for a while so feel free to change the PR or create a new one.

**Comment by jonathan-teh** [comment]
This can be closed as completed, it shipped in 7.00.
